File: extension/core/src/main/java/org/camunda/bpm/extension/process_test_coverage/engine/ElementCoverageParseListener.java
/*
 * Copyright 2020 FlowSquad GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package org.camunda.bpm.extension.process_test_coverage.engine;

import org.camunda.bpm.engine.delegate.DelegateExecution;
import org.camunda.bpm.engine.delegate.ExecutionListener;
import org.camunda.bpm.engine.impl.bpmn.parser.AbstractBpmnParseListener;
import org.camunda.bpm.engine.impl.pvm.process.ActivityImpl;
import org.camunda.bpm.engine.impl.pvm.process.ScopeImpl;
import org.camunda.bpm.engine.impl.pvm.process.TransitionImpl;
import org.camunda.bpm.engine.impl.util.xml.Element;
import org.camunda.bpm.extension.process_test_coverage.model.Collector;
import org.camunda.bpm.extension.process_test_coverage.model.Event;
import org.camunda.bpm.extension.process_test_coverage.model.EventSource;
import org.camunda.bpm.extension.process_test_coverage.model.EventType;

import java.util.Arrays;
import java.util.List;

import static org.camunda.bpm.engine.delegate.ExecutionListener.*;

/**
 * Parse listener for adding coverage information to the test state.
 */
public class ElementCoverageParseListener extends AbstractBpmnParseListener {

    /**
     * The collector of the currently running test.
     */
    private Collector coverageState;

    private static final List<String> EXECUTION_EVENTS = Arrays.asList(
            EVENTNAME_START,
            EVENTNAME_END);

    private final ExecutionListener executionListener;

    public ElementCoverageParseListener() {
        this.executionListener = this::execute;
    }

    private void execute(final DelegateExecution execution) {

        if (EVENTNAME_START.equals(execution.getEventName())) {
            final Event event = this.createEvent(
                    execution,
                    execution.getCurrentActivityId(),
                    EventSource.FlOW_NODE,
                    EventType.START);
            this.coverageState.addEvent(event);
        }

        if (EVENTNAME_END.equals(execution.getEventName())) {
            final Event event = this.createEvent(
                    execution,
                    execution.getCurrentActivityId(),
                    EventSource.FlOW_NODE,
                    EventType.END);
            this.coverageState.addEvent(event);
        }

        if (EVENTNAME_TAKE.equals(execution.getEventName())) {
            final Event event = this.createEvent(
                    execution,
                    execution.getCurrentTransitionId(),
                    EventSource.SEQUENCE_FLOW,
                    EventType.TAKE);
            this.coverageState.addEvent(event);
        }
    }

    private Event createEvent(final DelegateExecution execution, final String activityId, final EventSource eventSource, final EventType eventType) {
        return new Event(
                eventSource,
                eventType,
                activityId,
                execution.getBpmnModelElementInstance().getElementType().getTypeName(),
                execution.getProcessInstance().getProcessDefinitionId(),
                this.getProcessKey(execution));
    }

    private String getProcessKey(final DelegateExecution delegateExecution) {
        return delegateExecution.getProcessEngineServices().getRepositoryService().getProcessDefinition(delegateExecution.getProcessDefinitionId()).getKey();
    }

    @Override
    public void parseUserTask(final Element userTaskElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseBoundaryErrorEventDefinition(final Element errorEventDefinition, final boolean interrupting, final ActivityImpl activity, final ActivityImpl nestedErrorEventActivity) {
        // this.addExecutionListener(activity);
    }

    @Override
    public void parseBoundaryEvent(final Element boundaryEventElement, final ScopeImpl scopeElement, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseBoundaryMessageEventDefinition(final Element element, final boolean interrupting, final ActivityImpl activity) {
        //this.addExecutionListener(activity);
    }

    @Override
    public void parseBoundarySignalEventDefinition(final Element signalEventDefinition, final boolean interrupting, final ActivityImpl activity) {
        //this.addExecutionListener(activity);
    }

    @Override
    public void parseBoundaryTimerEventDefinition(final Element timerEventDefinition, final boolean interrupting, final ActivityImpl activity) {
        //this.addExecutionListener(activity);
    }

    @Override
    public void parseBusinessRuleTask(final Element businessRuleTaskElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseCallActivity(final Element callActivityElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseCompensateEventDefinition(final Element compensateEventDefinition, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseEndEvent(final Element endEventElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseEventBasedGateway(final Element eventBasedGwElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseExclusiveGateway(final Element exclusiveGwElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseInclusiveGateway(final Element inclusiveGwElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseIntermediateCatchEvent(final Element intermediateEventElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseIntermediateMessageCatchEventDefinition(final Element messageEventDefinition, final ActivityImpl activity) {
        // this.addExecutionListener(activity);
    }

    @Override
    public void parseIntermediateSignalCatchEventDefinition(final Element signalEventDefinition, final ActivityImpl activity) {
        // this.addExecutionListener(activity);
    }

    @Override
    public void parseIntermediateThrowEvent(final Element intermediateEventElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseIntermediateTimerEventDefinition(final Element timerEventDefinition, final ActivityImpl activity) {
        // this.addExecutionListener(activity);
    }

    @Override
    public void parseManualTask(final Element manualTaskElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseMultiInstanceLoopCharacteristics(final Element activityElement, final Element multiInstanceLoopCharacteristicsElement, final ActivityImpl activity) {
        // DO NOT IMPLEMENT
        // we do not notify on entering a multi-instance activity, this will be done for every single execution inside that loop.
    }

    @Override
    public void parseParallelGateway(final Element parallelGwElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseReceiveTask(final Element receiveTaskElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseScriptTask(final Element scriptTaskElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseSendTask(final Element sendTaskElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseSequenceFlow(final Element sequenceFlowElement, final ScopeImpl scopeElement, final TransitionImpl transition) {
        this.addExecutionListener(transition);
    }

    @Override
    public void parseServiceTask(final Element serviceTaskElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseStartEvent(final Element startEventElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseSubProcess(final Element subProcessElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseTask(final Element taskElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    @Override
    public void parseTransaction(final Element transactionElement, final ScopeImpl scope, final ActivityImpl activity) {
        this.addExecutionListener(activity);
    }

    private void addExecutionListener(final ActivityImpl activity) {
        for (final String event : EXECUTION_EVENTS) {
            activity.addListener(event, this.executionListener);
        }
    }

    private void addExecutionListener(final TransitionImpl transition) {
        transition.addListener(EVENTNAME_TAKE, this.executionListener);
    }

    public void setCoverageState(final Collector coverageState) {
        this.coverageState = coverageState;
    }

}

File: extension/core/src/main/kotlin/org/camunda/bpm/extension/process_test_coverage/engine/ElementCoverageParseListener.kt
/*
 * Copyright 2020 FlowSquad GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.camunda.bpm.extension.process_test_coverage.engine

import org.camunda.bpm.engine.delegate.DelegateExecution
import org.camunda.bpm.engine.delegate.ExecutionListener
import org.camunda.bpm.engine.impl.bpmn.parser.AbstractBpmnParseListener
import org.camunda.bpm.engine.impl.pvm.process.ActivityImpl
import org.camunda.bpm.engine.impl.pvm.process.ScopeImpl
import org.camunda.bpm.engine.impl.pvm.process.TransitionImpl
import org.camunda.bpm.engine.impl.util.xml.Element
import org.camunda.bpm.extension.process_test_coverage.model.Collector
import org.camunda.bpm.extension.process_test_coverage.model.Event
import org.camunda.bpm.extension.process_test_coverage.model.EventSource
import org.camunda.bpm.extension.process_test_coverage.model.EventType
import java.time.Instant

/**
 * Parse listener for adding coverage information to the test state.
 */
class ElementCoverageParseListener : AbstractBpmnParseListener() {

    companion object {
        private val EXECUTION_EVENTS = listOf(
            ExecutionListener.EVENTNAME_START,
            ExecutionListener.EVENTNAME_END
        )
    }

    /**
     * The collector of the currently running test.
     */
    private lateinit var coverageState: Collector
    private val executionListener: ExecutionListener = ExecutionListener { execution: DelegateExecution -> execute(execution) }

    fun setCoverageState(coverageState: Collector) {
        this.coverageState = coverageState
    }


    private fun execute(execution: DelegateExecution) {
        require(this::coverageState.isInitialized) { "Coverage state must be initialized" }
        if (ExecutionListener.EVENTNAME_START == execution.eventName) {
            val event = createEvent(
                execution,
                execution.currentActivityId,
                EventSource.FLOW_NODE,
                EventType.START
            )
            coverageState.addEvent(event)
        }
        if (ExecutionListener.EVENTNAME_END == execution.eventName) {
            val event = createEvent(
                execution,
                execution.currentActivityId,
                EventSource.FLOW_NODE,
                EventType.END
            )
            coverageState.addEvent(event)
        }
        if (ExecutionListener.EVENTNAME_TAKE == execution.eventName) {
            val event = createEvent(
                execution,
                execution.currentTransitionId,
                EventSource.SEQUENCE_FLOW,
                EventType.TAKE
            )
            coverageState.addEvent(event)
        }
    }

    private fun createEvent(execution: DelegateExecution, activityId: String, eventSource: EventSource, eventType: EventType): Event {
        return Event(
            eventSource,
            eventType,
            activityId,
            execution.bpmnModelElementInstance.elementType.typeName,
            getProcessKey(execution),
            Instant.now().epochSecond
        )
    }

    private fun getProcessKey(delegateExecution: DelegateExecution): String {
        return delegateExecution.processEngineServices.repositoryService.getProcessDefinition(delegateExecution.processDefinitionId).key
    }

    override fun parseUserTask(userTaskElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseBoundaryErrorEventDefinition(errorEventDefinition: Element, interrupting: Boolean, activity: ActivityImpl, nestedErrorEventActivity: ActivityImpl) {
        // this.addExecutionListener(activity);
    }

    override fun parseBoundaryEvent(boundaryEventElement: Element, scopeElement: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseBoundaryMessageEventDefinition(element: Element, interrupting: Boolean, activity: ActivityImpl) {
        //this.addExecutionListener(activity);
    }

    override fun parseBoundarySignalEventDefinition(signalEventDefinition: Element, interrupting: Boolean, activity: ActivityImpl) {
        //this.addExecutionListener(activity);
    }

    override fun parseBoundaryTimerEventDefinition(timerEventDefinition: Element, interrupting: Boolean, activity: ActivityImpl) {
        //this.addExecutionListener(activity);
    }

    override fun parseBusinessRuleTask(businessRuleTaskElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseCallActivity(callActivityElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseCompensateEventDefinition(compensateEventDefinition: Element, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseEndEvent(endEventElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseEventBasedGateway(eventBasedGwElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseExclusiveGateway(exclusiveGwElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseInclusiveGateway(inclusiveGwElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseIntermediateCatchEvent(intermediateEventElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseIntermediateMessageCatchEventDefinition(messageEventDefinition: Element, activity: ActivityImpl) {
        // this.addExecutionListener(activity);
    }

    override fun parseIntermediateSignalCatchEventDefinition(signalEventDefinition: Element, activity: ActivityImpl) {
        // this.addExecutionListener(activity);
    }

    override fun parseIntermediateThrowEvent(intermediateEventElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseIntermediateTimerEventDefinition(timerEventDefinition: Element, activity: ActivityImpl) {
        // this.addExecutionListener(activity);
    }

    override fun parseManualTask(manualTaskElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseMultiInstanceLoopCharacteristics(activityElement: Element, multiInstanceLoopCharacteristicsElement: Element, activity: ActivityImpl) {
        // DO NOT IMPLEMENT
        // we do not notify on entering a multi-instance activity, this will be done for every single execution inside that loop.
    }

    override fun parseParallelGateway(parallelGwElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseReceiveTask(receiveTaskElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseScriptTask(scriptTaskElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseSendTask(sendTaskElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseSequenceFlow(sequenceFlowElement: Element, scopeElement: ScopeImpl, transition: TransitionImpl) {
        this.addExecutionListener(transition)
    }

    override fun parseServiceTask(serviceTaskElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseStartEvent(startEventElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseSubProcess(subProcessElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseTask(taskElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    override fun parseTransaction(transactionElement: Element, scope: ScopeImpl, activity: ActivityImpl) {
        this.addExecutionListener(activity)
    }

    private fun addExecutionListener(activity: ActivityImpl) {
        for (event in EXECUTION_EVENTS) {
            activity.addListener(event, executionListener)
        }
    }

    private fun addExecutionListener(transition: TransitionImpl) {
        transition.addListener(ExecutionListener.EVENTNAME_TAKE, executionListener)
    }

}